rustsec/src/cached_index.rs

```diff
@@ -36,47 +85,134 @@ impl CachedIndex {
     }
 
     /// Open the local crates.io index
+    ///
+    /// If this opens a git index, it allows reading of index entries from the
+    /// repository
+    ///
+    /// If this is a sparse index, it only allows reading of index entries that
+    /// are already cached locally
```

Comment:
This documentation confuses me. What happens if the crate I open is not in the cache?

Ideally there should be an API that transparently uses the local on-disk cache and fetches remote data only when necessary, where "when necessary" is determined in the same way as in Cargo itself.

Reply:
If a crate is not in the index cache it will return `None`.

Cargo will always do a remote operation unless in offline mode, as the only way to check if the local cache entry is up to date is to send the etag (or last-modified but AFAICT cargo only uses the e-tag) of the cache entry to either get an updated entry or a `NOT_MODIFIED`. The comment is just saying that it works the exact same as the git index in the sense that only local information will be used, the difference is that in the git case crates that haven't been previously queried by cargo/other tools _might_ still be present in the git blobs, and thus readable (if outdated), whereas in the sparse case they won't exist in the local cache.

Reply:
Got it. Thank you!